initial cuda core and bgsegm implementation

netbrain opened the pull request:
Initial implementation of cuda core and cudabgsegm with MOG and MOG2 algorithms

netbrain commented:
@deadprogram Hi again! started on implementing cuda. Im on thin ground here, so please take a look if it looks right to you. Cheers.

netbrain commented:
I guess we need to have cuda compiled on the CI systems :(

netbrain commented on go.mod:
```diff
@@ -1 +1,3 @@
 module gocv.io/x/gocv
+
+go 1.12
```
@deadprogram anyway to ignore the auto inclusion of the go version? Not sure what is adding this to this file, if it is go itself or my IDEA (goland).

netbrain replied:
https://youtrack.jetbrains.com/issue/GO-7121

deadprogram replied:
So, you will remove that line manually? Seems like that might be your own option at the moment.

netbrain commented:
@deadprogram looks like I need some input on how i can either exclude these cuda dependent files from compilation or add cuda support to travis and appveyor.

deadprogram commented:
The 2 options generally for this are either a build tag to exclude those files or else a sub-package.

In some cases, both, such as how the openvino sub-package is organized.

netbrain commented:
Tried a fast experiment where i created a new cuda package. However running the tests fails with `
cuda.h:5:10: fatal error: opencv2/opencv.hpp: No such file or directory
 #include <opencv2/opencv.hpp>
`
Im getting identical results when trying to run tests in the openvino package. I also tried with a custom environment but that seemed to create a whole lot of other errors.

Any idea what's going on? Ill look more into it later.

netbrain commented:
@deadprogram ping, just a friendly reminder :)

deadprogram commented:
Hi @netbrain this is an amazing PR. I have not had a chance to install the dependencies and test on my machine yet. 

As you have already identified, cuda would need to be installed on Travis and Appveyor in order to run the tests on the CI servers. However the needed access to physical hardware on those services does not exist. So that will not be happening anytime soon.

For those people who want to take the plunge, there should be a `cuda/README.md` file, describing what you need to do to install cuda on the various supported platforms. Also what version(s) of cuda are supported.

What do you think? Anyone else want to chime in here?

netbrain commented:
Thank you @deadprogram, for this first PR i would simply want to create the base platform for cuda features. As for the cuda/README file, sure i can add that, at least a minimal one. As you can see in the PR i have added a make target for installing opencv with cuda support, but i guess it should be possible to create a similar script that actually installs cuda itself aswell. But at the same time I really dont want to do too much in this PR, and instead add the missing pieces over time.

So to summarize, add a cuda/README file that guides the user to the cuda installation page (https://docs.nvidia.com/cuda/index.html) and document the cuda makefile target? 

Will that be sufficient for this PR?

deadprogram commented:
That sounds like a very reasonable plan to me.

netbrain commented:
@deadprogram  Added the readme. I think this should be good as a starting point for cuda contributions. I also refactored the names for some of the methods. They were prefixed with "Cuda" but i removed that as they now live in their own package.

netbrain commented:
@deadprogram so what is the plan with this? does it need more work?

deadprogram commented:
I think this is an incredible start for the community to be able to contribute to. Now merging, thank you for the great work!
